### PR TITLE
Fix jsonrpc tests

### PR DESCRIPTION
**Pull request by gsalgado**
Also moves the logic to keep track of the current head candidate into ChainService

There are a couple XXX comments for changes that I'm not sure are correct, although they were necessary to make the tests pass

Depends on https://github.com/ethereum/pyethereum/pull/736

### CODE REVIEW

**Comment by gsalgado** [comment]
It'd be great if you could review this one, @janx :-)

File: pyethapp/eth_service.py [hunk 1]
```diff
@@ -168,11 +159,15 @@ def __init__(self, app):
         coinbase = app.services.accounts.coinbase
         env = Env(self.db, sce['block'])
 
-        genesis_data = sce['genesis_data'] if 'genesis_data' in sce else dict()
-        self.chain = Chain(env=env,
-                           genesis=genesis_data,
-                           coinbase=coinbase,
-                           new_head_cb=self._on_new_head)
+        genesis_data = mk_genesis_data(env)
```
Comment (janx) [inline pyethapp/eth_service.py]: I think we need to keep `sec['genesis_data']`, which is generated from configurations and cmd line params.

File: pyethapp/eth_service.py [hunk 2]
```diff
@@ -168,11 +159,15 @@ def __init__(self, app):
...
+        # XXX: Workaround for
+        # https://github.com/ethereum/pyethereum/issues/726
+        genesis_data["extraData"] = genesis_data["extra_data"]
+        genesis_data["parentHash"] = genesis_data["prevhash"]
+        genesis_data["gasLimit"] = genesis_data["gas_limit"]
```
Comment (janx) [inline pyethapp/eth_service.py]: `genesis_data` should be constructed and passed in.

File: pyethapp/eth_service.py [hunk 3]
```diff
@@ -206,11 +206,34 @@ def is_mining(self):
             return self.app.services.validator.active
         return False
 
+    def getReceipts(self, block):
```
Comment (janx) [inline pyethapp/eth_service.py]: I suggest moving get receipts abstraction into pyethereum. Notice pyethereum is under big refactoring, new changes should base on develop_with_state_revamp branch.

Method name should following underscore naming convention.

File: pyethapp/eth_service.py (hunk 2 above)
Comment (gsalgado) [reply pyethapp/eth_service.py]: Would it be ok to have it constructed here, for convenience, when it's not passed in?

File: pyethapp/eth_service.py [hunk 4]
```diff
@@ -169,11 +158,17 @@ def __init__(self, app):
...
-                           coinbase=coinbase,
-                           new_head_cb=self._on_new_head)
+        genesis_data = sce.get('genesis_data', {})
+        if not genesis_data:
+            genesis_data = mk_genesis_data(env)
+            # XXX: Workaround for
+            # https://github.com/ethereum/pyethereum/issues/726
+            genesis_data["extraData"] = genesis_data["extra_data"]
+            genesis_data["parentHash"] = genesis_data["prevhash"]
+            genesis_data["gasLimit"] = genesis_data["gas_limit"]
```
Comment (janx) [inline pyethapp/eth_service.py]: Do you think we can fix https://github.com/ethereum/pyethereum/blob/develop/ethereum/genesis_helpers.py#L81 to generate camelCase keys and remove L164-L168?

Comment (gsalgado) [reply pyethapp/eth_service.py]: Sure, I'll do that

**Comment by AndreMiras** [comment]
How nice! I was about to pickup some of these issues. Well done guys!

**Pull request merged**